### PR TITLE
Make SyncDroplessArena allocations thread-local in the fast path

### PR DESCRIPTION
**Pull request by Zoxc**
This works by having a thread local chunk to allocate from and a global list of chunks for `in_arena` to use.

r? @eddyb

### CODE REVIEW

**Comment by eddyb** [comment]
I don't know how any of this concurrent code works, sorry! (Or rather, why it might be safe)
Maybe if @aturon gets back, he can help with reviewing it?

cc @rust-lang/compiler Is anyone else comfortable with concurrent data structures and willing to review?

**Comment by nikomatsakis** [comment]
I'll review it and/or get @aturon to do so =)

**Comment by Dylan-DPC** [comment]
ping from triage @nikomatsakis @aturon any updates on this?

**Comment by aturon** [comment]
Hey @Dylan-DPC, I plan to review this one, however at the moment I'm getting up to speed on the  concurrency approach in general.

**Comment by Dylan-DPC** [comment]
thanks. No issues.

**Comment by eddyb** [comment]
Also pinging @Gankro and @stjepang, in case they want to review as well.

**Comment by Gankro** [comment]
I do not

**Comment by hdhoang** [comment]
ping from triage @stjepang, could you provide comment on this PR?

File: src/libarena/lib.rs [hunk 1]
```diff
@@ -555,64 +544,164 @@ impl DroplessArena {
...
+        self.ptr.set(final_address as *mut T);
+        assert!(self.ptr <= self.end);
     }
 
+    /// Grows the arena.
     #[inline(always)]
-    pub fn clear(&mut self) {
-        self.lock.get_mut().clear();
+    fn grow(&self, n: usize, chunks: &mut Vec<TypedArenaChunk<T>>) {
+        unsafe {
```
Comment (Centril) [inline src/libarena/lib.rs]: Would be great if these large `unsafe { .. }` blocks would have comments justifying their soundness.
(Applies throughout the PR.)

**Comment by ProgrammaticNajel** [comment]
Ping from triage. @aturon any updates on this? Thanks

**Comment by JohnCSimon** [comment]
Pinging again from triage
This PR has sat idle waiting for review for eight days. 
@Centril @aturon  

Thanks.

**Comment by Dylan-DPC** [comment]
Waiting for a review from @rust-lang/compilers

File: src/libarena/lib.rs [hunk 2]
```diff
@@ -555,64 +544,164 @@ impl DroplessArena {
...
+            let (chunk, mut new_capacity);
+            if let Some(last_chunk) = chunks.last_mut() {
```
Comment (oli-obk) [inline src/libarena/lib.rs]: The statements in the `Some` block of this `if let` deserve some documenting comments.

File: src/libarena/lib.rs [hunk 3]
```diff
@@ -555,64 +544,164 @@ impl DroplessArena {
...
-        // Extend the lifetime of the result since it's limited to the lock guard
-        unsafe { &mut *(self.lock.lock().alloc_raw(bytes, align) as *mut [u8]) }
+        unsafe {
+            assert!(bytes != 0);
+
+            let current = &*self.current;
+
+            current.align(align);
+
+            let future_end = intrinsics::arith_offset(current.ptr.get(), bytes as isize);
```
Comment (oli-obk) [inline src/libarena/lib.rs]: why not use https://doc.rust-lang.org/std/primitive.pointer.html#method.wrapping_offset here? Intrinsics should not be used except in their wrapper code

Comment (oli-obk) [reply src/libarena/lib.rs]: Alternatively you can use `offset_from` to compute the offset between `end` and `ptr` and see if that is bigger than `bytes`. I think I'd prefer it that way around.

File: src/libarena/lib.rs [hunk 4]
```diff
@@ -555,64 +544,164 @@ impl DroplessArena {
...
+            if (future_end as *mut u8) >= current.end.get() {
+                self.grow(bytes);
+            }
+
+            let ptr = current.ptr.get();
+            // Set the pointer past ourselves
+            current.ptr.set(
+                intrinsics::arith_offset(current.ptr.get(), bytes as isize) as *mut u8,
```
Comment (oli-obk) [inline src/libarena/lib.rs]: since we just grew the thing, we can now use `ptr::offset`, right?

**Review by oli-obk — CHANGES_REQUESTED** [review]

**Comment by JohnCSimon** [comment]
Ping from triage - 
@Zoxc @oli-obk @rust-lang/compilers
This has sat idle for two weeks. What needs to happen for this pr?

Thanks

**Comment by joelpalmer** [comment]
Ping from Triage: closing due to inactivity. Please re-open if there are updates. @Zoxc

File: src/libarena/lib.rs (hunk 3 above)
Comment (Zoxc) [reply src/libarena/lib.rs]: Oh, we stabilized this intrinsic? That seems unfortunate as it can be used to create invalid pointers (which are UB in C).

Comment (oli-obk) [reply src/libarena/lib.rs]: It's not stable. and it's still unsafe, so if you think there are problems with it, please raise them on the tracking issue (https://github.com/rust-lang/rust/issues/41079). `wrapping_offset_from` is safe.

**Pull request closed without merging**